### PR TITLE
[srp-client] add `Srp::Client::ClearService()`

### PR DESCRIPTION
**Pull request by abtink**
This method clears a service from SRP client, immediately removing it
from the list. Unlike `RemoveService()` which sends an update message
to server to remove/unregister the service, this method clears the
service from client list without any interaction with server. On a
successful call to this method, the `Callback` will NOT be called and
service entry can be reclaimed and re-used by the caller immediately.

This method can be used along with a subsequent call to `AddService()`
(potentially reusing the same service entry with the same service and
instance names) to update some of the parameters in an existing
service.

This commit adds `otSrpClientClearService()` as a related public OT
API and support for this in CLI (command `service delete`). It also
adds support for the deleting service in `NcpBase` and spinel.

### CODE REVIEW

File: include/openthread/srp_client.h [hunk 1]
```diff
@@ -473,6 +473,26 @@ otError otSrpClientAddService(otInstance *aInstance, otSrpClientService *aServic
...
+ * aService entry with the same service and instance names) to update some of the parameters in an existing service.
+ *
+ * @param[in] aInstance        A pointer to the OpenThread instance.
+ * @param[in] aService         A pointer to a `otSrpClientService` instance to delete.
+ *
+ * @retval OT_ERROR_NONE       The @p aService is deleted successfully. It can be reclaimed and re-used immediately.
+ * @retval OT_ERROR_NOT_FOUND  The service could not be found in the list.
+ *
+ */
+otError otSrpClientDeleteService(otInstance *aInstance, otSrpClientService *aService);
```
Comment (simonlingoogle) [inline include/openthread/srp_client.h:494]: I think this API is quite confusing and error-prone:
- User may call this method to delete a service and expect the service to be removed on server (without reading the Doxygen carefully).
- Users generally don't need an API to delete a service without deregistering from the server. So, this API should not exist.

If we want to support updating existing service, maybe we can allow user to modify the `otSrpClientService` buffer in place and call an API (e.g. otSrpClientCommitService) to commit the change.

Thoughts?

Comment (abtink) [reply include/openthread/srp_client.h:494]: > If we want to support updating existing service, maybe we can allow user to modify the otSrpClientService buffer in place and call an API (e.g. otSrpClientCommitService) to commit the change.'

I think allowing caller to change an `otSrpClientService` which it is does not own anymore (has passed it to OT and is now owned by OT core and SRP client) is more error-prone (easier to use incorrectly and can lead to tricky racing bugs). I describe the reason why we do not want to allow this in https://github.com/openthread/openthread/issues/6586#issuecomment-834774771 (copying the relevant part below): 

> Back to the point about allowing an existing/registered `otSrpClientService` instances to be changed by caller (after it is given to the OT):
> - I agree this can help allow reusing the same entry and we can add a new OT API to "update" an existing service (not need to remove and add with two different entry instances).
> - My concern is that if we relax the requirement for `otSrpClientService` to not be changed by caller, it may lead to potential incorrect use and/or tricky bugs.
>    -  For example, we can have a race situation if the next layer code tries to change some fields/pointers in an already registered `otSrpClientService` struct from a different OS context (e.g., different RTOS task) while the OT core is also accessing the same content.


And more in https://github.com/openthread/openthread/issues/6586#issuecomment-835925083:

> Thinking more on this, I came up with an approach to allow caller (next layer code) to re-use an existing service entry to update some service paramter without relaxing our requirement for the `otSrpClientService` not to be changed while it is given to and owned by OT core (`Srp::Client` module).
> 
> The idea is we add a new `DeleteService()` (or `otSrpClientDeleteService()` as public OT API) method
> - This method deletes a service from SRP client, immediately removing it from the list. 
> - Unlike `RemoveService()` which sends an SRP update message to server to remove/unregister the service on server , this method deletes the service from the client list without any interaction with server. 
> - On a successful call to this method, the `Callback` will NOT be called and service entry can be reclaimed and re-used by the caller immediately.
> - Basically after `DeleteService()` the service is removed on the client side, but may still be present on the server till it expires its current lease time.

------

I don't think it will be hard for caller/user to distinguish between `RemoveServcie` and `DeleteService`. We already have `RemoveHostAndService` and `ClearHostAndService` which mirror similar behavior (one informs server, the other one clears everything immediately with no interaction with server).

File: src/lib/spinel/spinel.h [hunk 2]
```diff
@@ -4112,15 +4112,26 @@ enum
...
      *   `S` : The service weight.
      *
-     * During remove operation, only service name and service instance name would be used.
+     * For `SPINEL_CMD_PROP_VALUE_REMOVE` command, the following format is used:
+     *
+     *   `U` : The service name labels (e.g., "_chip._udp", not the full domain name.
+     *   `U` : The service instance name label (not the full name).
+     *   `b` : Indicates whether to delete the service entry (optional).
+     *
+     * The last boolean (`b`) field is optional. When included it indicates on `true` to delete the service (deleter it
```
Comment (Irving-cl) [inline src/lib/spinel/spinel.h:4131]:
```suggestion
     * The last boolean (`b`) field is optional. When included it indicates on `true` to delete the service (delete it
```

**Review by Irving-cl — APPROVED** [review]
LGTM 👍

File: include/openthread/srp_client.h (hunk 1 above)
Comment (simonlingoogle) [reply include/openthread/srp_client.h:494]: > For example, we can have a race situation if the next layer code tries to change some fields/pointers in an already registered otSrpClientService struct from a different OS context (e.g., different RTOS task) while the OT core is also accessing the same content.

User should already be aware that OT APIs should be called in the OT thread. Furthermore, we can make it clear in the doxygen of `otSrpServiceCommitService` interface.

> Unlike RemoveService() which sends an SRP update message to server to remove/unregister the service on server , this method deletes the service from the client list without any interaction with server. 

The problem is that we are introducing an API `DeleteService`, which however is **NOT** meant to be used to **delete** a service. 

> We already have RemoveHostAndService and ClearHostAndService which mirror similar behavior (one informs server, the other one clears everything immediately with no interaction with server).

Even though we have `otSrpClientClearHostAndServices`, we actually never use it. I think it more logical for user to call `otSrpClientClearHostAndServices` to clear the host and all services on it in the case when the SRP client knows somehow that the SRP server is down. But for `otSrpClientDeleteService`, I can't think of any use case that user may want to delete one service without notifying the SRP server.

---

If we want to provide an API for updating an existing service, I think we should provide an API just for that, and make it clear how user can update the service data. We have already decided in our original design that it's callers' responsibility to manage the `otSrpClientService` buffers, so it should be fine if we continue to follow this design?

Thoughts?

Comment (abtink) [reply include/openthread/srp_client.h:494]: Thanks @simonlingoogle  for your input, but I don't agree with you on this  and feel strongly (and know from my experience) that _"relaxing the requirement for the `otSrpClientService` to be immutable after ownership is passed to OT"_ is not a good idea and it is error-prone (can lead to tricky bugs and incorrect API use).

> User should already be aware that OT APIs should be called in the OT thread. 

The `otSrpClientServcie` is not an API. It is an object which we expect the caller to allocate it, populate it, and then pass it to OT API. It can be created/populated in another OS thread/task (any OS context). There is no reason to restrict how it is being created by caller in the first place (in general sense) and limit it to same OS thread/task as OT. Once it is given to OT, it is owned by OT and becomes immutable.

> Furthermore, we can make it clear in the doxygen of otSrpServiceCommitService interface.

It is much simpler and easier for developers to understand and follow a rule that once ownership is given to OT the `otSrpClientServcie` become immutable (must not be changed by caller anymore). The ownership model is commonly used in multi-threaded (multi-task) designs.

Compared to something like this:
- `otSrpClientServcie` can be allocated and updated anyway the user want (in any thread/task OS context) before it is passed to OT APIs. 
- But after it is passed to OT APIs, it can be changed/updated only from same OS context where the OT API are being called.

This is complex to explain and hard to follow for users (easy to forget and use incorrectly).

> The problem is that we are introducing an API DeleteService, which however is NOT meant to be used to delete a service.

The `DeleteServcie()` is only meant for deleting a service (without notifying the server) and claiming back ownership of `otSrpClientServcie` immediately.

The caller can choose to use it to update a service by deleting it, claiming back the entry and then adding it back sameway that `RemoveServcie()` can be used today for this.

> If we want to provide an API for updating an existing service, I think we should provide an API just for that, and make it clear how user can update the service data. We have already decided in our original design that it's callers' responsibility to manage the otSrpClientService buffers, so it should be fine if we continue to follow this design?

We don't want to provide an API for updating an existing servcie, because it relaxes the notion of ownership of service entries.

Comment (simonlingoogle) [reply include/openthread/srp_client.h:494]: > The `DeleteServcie()` is only meant for deleting a service (without notifying the server) and claiming back ownership of `otSrpClientServcie` immediately.
> 
> The caller can choose to use it to update a service by deleting it, claiming back the entry and then adding it back sameway that `RemoveServcie()` can be used today for this.

As I have said, `deleting a service without notifying the server` is almost never what users want to do. And `Delete` is very similar to `Remove`, causing confusion. 

Should we use another name like `otSrpClientDetachService` to make it clear that this API just detaches the `otSrpClientServcie` object from the SRP client, without doing anything else?

Comment (abtink) [reply include/openthread/srp_client.h:494]: > As I have said, deleting a service without notifying the server is almost never what users want to do

We actually mention one possible use-case of "delete" API in the documentation of the API itself, i.e. if the API users do not want to allocate a new entry to update an existing service, they can use "delete" to reclaim an existing service entry immediately and then to use it again to add. There may be other use-cases in future.

> And Delete is very similar to Remove, causing confusion.

I think we can expect API user to read documentation and understand the behavior. 

Besides we already have other APIs with similar behavior, `otSrpClientRemoveHostAndServices()` and 
`otSrpClientClearHostAndServices()`.

I initially actually used `ClearService()` instead of `DeleteService()` for the new method (it may be evident from name of the branch `srp-clinet/clear-service-method` 😄) but I felt `DeleteService()` was better. I am not sure if `DetachServcie()` name would be more clear.  I don't mind going back to `ClearService()` if other agree.

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

Even though the method is a little confusing, it solves the problem we have at hand.

File: include/openthread/srp_client.h (hunk 1 above)
Comment (wgtdkp) [reply include/openthread/srp_client.h:494]: @abtink I think it make sense to use the similar name with `otSrpClientClearHostAndServices` to help user use the API correctly. That said, I lean to `ClearService` even that `clear` usually means remove a bunch of things. But considering that a service includes a few RRs (PTR, SRV, TXT) and buffers, I think it should not be very strange to call it `clear` 😄 .

**Review by wgtdkp — APPROVED** [review]
LGTM 👍

File: include/openthread/srp_client.h (hunk 1 above)
Comment (abtink) [reply include/openthread/srp_client.h]: In new pushed commit, renamed it back to` ClearService()`.

File: include/openthread/srp_client.h [hunk 3]
```diff
@@ -473,6 +473,26 @@ otError otSrpClientAddService(otInstance *aInstance, otSrpClientService *aServic
  */
 otError otSrpClientRemoveService(otInstance *aInstance, otSrpClientService *aService);
 
+/**
+ * This function clears a service, immediately removing it from the client service list.
+ *
+ * Unlike `otSrpClientRemoveService()` which sends an update message to server to remove/unregister the service, this
```
Comment (jwhui) [inline include/openthread/srp_client.h:479]:
```suggestion
 * Unlike `otSrpClientRemoveService()` which sends an update message to the server to remove the service, this
```

File: src/core/net/srp_client.hpp [hunk 4]
```diff
@@ -499,6 +499,22 @@ class Client : public InstanceLocator, private NonCopyable
 
     Error RemoveService(Service &aService);
 
+    /**
+     * This method clears a service, immediately removing it from the client service list.
+     *
+     * Unlike `RemoveService()` which sends an update message to server to remove/unregister the service, this
```
Comment (jwhui) [inline src/core/net/srp_client.hpp:505]:
```suggestion
     * Unlike `RemoveService()` which sends an update message to the server to remove the service, this
```

File: src/lib/spinel/spinel.h [hunk 5]
```diff
@@ -4112,15 +4112,26 @@ enum
...
+     * For `SPINEL_CMD_PROP_VALUE_REMOVE` command, the following format is used:
+     *
+     *   `U` : The service name labels (e.g., "_chip._udp", not the full domain name.
+     *   `U` : The service instance name label (not the full name).
+     *   `b` : Indicates whether to clear the service entry (optional).
+     *
+     * The last boolean (`b`) field is optional. When included it indicates on `true` to clear the service (clear it
+     * on client immediately with no interaction to server) and on `false` to remove the service (inform server and
+     * wait for the service entry to be removed on server). If it is not specified service will be removed (informing
+     * the server).
```
Comment (jwhui) [inline src/lib/spinel/spinel.h:4134]:
```suggestion
     * wait for the service entry to be removed on server). If it is not included, the value is `false`.
```

**Review by jwhui — APPROVED** [review]
